### PR TITLE
initial ts documentation

### PR DESCRIPTION
**Pull request by evanshortiss**
This should serve as a start for TypeScript documentation. We could link to it in a PR/Issue template?

### CODE REVIEW

File: docs/TypeScript.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+<h1 align="center">Fastify</h1>
+
+## TypeScript
+Fastify is shipped with the a typings file so it should "just work" when your used in a TypeScript application.
```
Comment (jsumners) [inline docs/TypeScript.md]: `should "just work" when your used in a TypeScript application.` => `should "just work" when used in a TypeScript application.`

File: docs/TypeScript.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+If you're using plugins be sure to read the sections below.
```
Comment (jsumners) [inline docs/TypeScript.md]: How about this instead?: `Plugins may or may not include typings. See [Plugin Types](#plugin-types) for more information.`

File: docs/TypeScript.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
+<h1 align="center">Fastify</h1>
+
+## TypeScript
```
Comment (jsumners) [inline docs/TypeScript.md]: Add `<a id="typescript"></a>` prior to this line.

File: docs/TypeScript.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  }
+}
+
+server.get('/ping', opts, (req, reply) => {
+  console.log(reply.res) // this is the http.ServerResponse with correct typings!
+  reply.code(200).send({ pong: 'it worked!' })
+})
+```
+
+## Contributing
```
Comment (jsumners) [inline docs/TypeScript.md]: Add `<a id="contributing"></a>` prior to this line.

File: docs/TypeScript.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+TypeScript related changes can be considered to fall into one of two categories:
+
+* Core - The typings bundled with fastify
+* Plugins - Fastify ecosystem plugins
+
+Make sure to read our `CONTRIBUTING.md` file before getting started to make sure things go smoothly!
+
+### Core Types
```
Comment (jsumners) [inline docs/TypeScript.md]: Add `<a id="core-types"></a>` prior to this line.

File: docs/TypeScript.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+When updating core types you should make a PR to this repository. Ensure you:
+
+1. Update `examples/typescript-server.ts` to reflect the changes (if necessary)
+2. Update `test/types/index.ts` to validate changes work as expected
+
+### Plugin Types
```
Comment (jsumners) [inline docs/TypeScript.md]: Add `<a id="plugin-types"></a>` prior to this line.

File: docs/TypeScript.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+Typings for plugins are hosted in DefinitelyTyped. This means when using plugins you should install like so:
+
+```
+npm install fastify-url-data @types/fastify-url-data
+```
+
+After this you should be good to go. Some types might not be available yet, so don't be shy about contributing.
+
+### Authouring Plugin Types
```
Comment (jsumners) [inline docs/TypeScript.md]: Add `<a id="authoring-plugin-types"></a>` prior to this line.

File: docs/TypeScript.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
+<h1 align="center">Fastify</h1>
+
+## TypeScript
+Fastify is shipped with the a typings file so it should "just work" when your used in a TypeScript application.
+
+If you're using plugins be sure to read the sections below.
+
+Here's a TypeScript example app. It closely aligns with the JavaScript examples:
```
Comment (jsumners) [inline docs/TypeScript.md]: `This example TypeScript app closely aligns with the JavaScript examples:`

**Comment by jsumners** [comment]
Is it not possible to add some description to the core typings that allows for unknown methods/properties on the core objects?

**Comment by evanshortiss** [comment]
@jsumners will action those ASAP. Travelling today 👍 

> Is it not possible to add some description to the core typings that allows for unknown methods/properties on the core objects?

Do you mean such that a user could just do:

```
server.blah()
```

If they happened to need a quick way to do that? It's possible in TS to do the following if you're super sure it'll be ok:

```
(server as any).blah()
```

Or even better, you can extend the server interface in your own code. [Here's a sample](https://github.com/evanshortiss/obd-parser-cli/blob/master/lib/interface.ts#L5) of doing so for a `commander` CLI I wrote a while back.

**Comment by jsumners** [comment]
I suggesting that the core declarations should allow for arbitrary methods/properties so that there doesn't _have_ to be typings added for every plugin/decoration to ever be written for Fastify.

**Comment by evanshortiss** [comment]
I don't think that's an issue core can solve. It's an overall TS language thing where properties and interfaces need to be correctly defined. You should play around and get a feel for it.

We could add:

```
[key: string]: any
```

Then developers could do:

```
server.a()
server.a()
```

**Comment by evanshortiss** [comment]
server.a()

**Comment by jsumners** [comment]
Why did you close this? Accident?

**Comment by evanshortiss** [comment]
Accidentally fat fingered. Reopening.

**Comment by delvedor** [comment]
@evanshortiss can you address @jsumners's suggestions?
Then we can merge this :)

**Comment by evanshortiss** [comment]
👍 think we're good @delvedor @jsumners? I added one more section related to http prototypes

**Comment by jsumners** [comment]
👍

**Pull request merged**